modelforge/potential/neighbors.py

```diff
@@ -0,0 +1,418 @@
+
+    This is compatible with TorchScript.
+
+
+    """
+
+    def __init__(
+        self,
+        cutoff: float,
+        displacement_function: OrthogonalDisplacementFunction,
+        only_unique_pairs: bool = False,
+    ):
+        """
+        Compute neighbor lists for inference, filtering pairs based on a cutoff distance.
+
+        Parameters
+        ----------
+        cutoff : float
+            The cutoff distance for neighbor list calculations.
+        displacement_function : OrthogonalDisplacementFunction
+            The function to calculate displacement vectors and distances between atom pairs, taking into account
+            the specified boundary conditions.
+        only_unique_pairs : bool, optional
+            Whether to only use unique pairs in the pair list calculation, by default False.
+        """
+
+        super().__init__()
+
+        self.register_buffer("cutoff", torch.tensor(cutoff))
+        self.register_buffer("only_unique_pairs", torch.tensor(only_unique_pairs))
+        self.displacement_function = displacement_function
+
+        self.indices = torch.tensor([])
+        self.i_final_pairs = torch.tensor([])
+        self.j_final_pairs = torch.tensor([])
+        log.info("Initializing Brute Force N^2 Neighborlist")
+
+    def forward(self, data: NNPInputTuple):
+        """
+        Prepares the input tensors for passing to the model.
+
+        This method handles general input manipulation, such as calculating
+        distances and generating the pair list. It also calls the model-specific
+        input preparation.
+
+        Parameters
+        ----------
+        data : NNPInputTuple
+            The input data provided by the dataset, containing atomic numbers,
+            positions, and other necessary information.
+
+        Returns
+        -------
+        PairListOutputs
+            Contains pair indices, distances (d_ij), and displacement vectors (r_ij) for atom pairs within the cutoff.
+        """
+        # ---------------------------
+        # general input manipulation
+        positions = data.positions
+        atomic_subsystem_indices = data.atomic_subsystem_indices
+
+        n = atomic_subsystem_indices.size(0)
+
+        # avoid reinitializing indices if they are already set and haven't changed
+        if self.indices.shape[0] != n:
+            # Generate a range of indices from 0 to n-1
+            self.indices = torch.arange(n, device=atomic_subsystem_indices.device)
+
+            # Create a meshgrid of indices
+            self.i_final_pairs, self.j_final_pairs = torch.meshgrid(
+                self.indices, self.indices, indexing="ij"
+            )
+            # We will only consider unique pairs; for non-unique pairs we can just appropriately copy
+            # the data as it will be faster than extra computations.
+            mask = self.i_final_pairs < self.j_final_pairs
```

Comment:
I think I might be missing something here. `torch.meshgrid` creates the N^2 (btw, `torch.cartesian_prod` seems to do the same operation but returns it in the list view that we want. did you have a chance to look at this?), which, for our purpose, are all possible interactions (plus the self interaction on the diagonal). Would it not be easier not to remove the lower triangle and regenerate it subsequently, but simply mask the diagonal? 

The used mask:
````            mask = self.i_final_pairs < self.j_final_pairs```
can still be applied when only unique paris are required

Reply:
And, I guess this is something we only need to do once, right? we can keep the N^2 matrix of atom indices in memory?

Reply:
We don't store the full N^2 matrix in memory after the initial calculation;  in the two lines below this we mask and save only the subset (which is the upper triangular removing the diagonal).  The Cartesian product basically doesn't the same thing as mesh grid but returns shape (N,2), not (2,N) which is what we want.  So I'm not sure what advantage this would be.

Reply:
I'll note we only generate once and just save this, unless the number of particles changes (in which case we need to regenerate).

Reply:
> We don't store the full N^2 matrix in memory after the initial calculation; in the two lines below this we mask and save only the subset (which is the upper triangular removing the diagonal).

Yes, I understand. But I was wondering if it's not better to store the N^2 matrix in memory. The upper triangular is only a special case needed for ANI, and can be easily recovered from the N^2 matrix. Therefore I am wondering if it is not easier and clearer to operate on the N^2 matri (and only removing the diagonal here)

Reply:
you clarifyied your seasons below, I consider this as resolved!